### PR TITLE
Add TPE Optimiser

### PR DESCRIPTION
**Pull request by JonnoFTW**
* Refactored out code
* Implemented TPE Optimiser from hyperopt
* Fixed spelling errors

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,20 +10,21 @@ Read more [here](https://github.com/fmfn/BayesianOptimization), [here](http://ga
...
-This information is used to sample more set of weights `--n_iters` number of times. This time we don't sample all of them in one go. Instead we sample once, merge the models,
+This information is used to sample more set of weights `--n_iters` number of times. This time we don't sample all of them in one go. Instead, we sample once, merge the models,
 generate and score the images and update the optimiser knowledge about the merging space. This way the optimiser can adapt the strategy step-by-step.
 
 At the end of the exploitation phase, the set of weights scoring the highest score are deemed to be the optimal ones.
 
 ## Juicy features
 
 - wildcards support
+- TPE or Bayesian Optimisers
```
Comment (s1dlx) [inline README.md:23]: here we can add Bergstra paper reference, it actually compares the two methods

File: sd_webui_bayesian_merger/tpe_optimiser.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+from hyperopt import Trials, hp, fmin, tpe, STATUS_OK
+
+
+class TPEOptimiser(Optimiser):
+
+    def _target_function(self, params):
+        res = self.sd_target_function(**params)
+        return {
+            'loss': -res,
```
Comment (s1dlx) [inline sd_webui_bayesian_merger/tpe_optimiser.py:15]: `-res` or `1/(res + 1e-10)` ?

Comment (JonnoFTW) [reply sd_webui_bayesian_merger/tpe_optimiser.py:15]: We want to maximise the value don't we? At least that's what I gathered from the Bayesian Optimiser usage

Comment (s1dlx) [reply sd_webui_bayesian_merger/tpe_optimiser.py]: yes indeed

I was wondering if there's a difference for tpe in minimizing a negative number towards -infinite or a number between 0 and 1

probably it doesn't matter

File: sd_webui_bayesian_merger/tpe_optimiser.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        best = self.trials.best_trial
+        print("Best:", best)
+
+        img_path = Path("logs", f"{self.merger.output_file.stem}.png")
+
+        convergence_plot(scores, figname=img_path)
+
+        unet_path = Path("logs", f"{self.merger.output_file.stem}-unet.png")
+        draw_unet(
+            best['params']['base_alpha'],
```
Comment (s1dlx) [inline sd_webui_bayesian_merger/tpe_optimiser.py:58]: this raises a `KeyError: 'params'`

perhaps that should be `best['spec']['base_alpha']`?

as for 
https://github.com/hyperopt/hyperopt/blob/6c5b33be93a71a4ae6a38f04a29b18020b78b702/hyperopt/base.py#L5-L12
```
- Trials - a list of documents including at least sub-documents:
    ['spec'] - the specification of hyper-parameters for a job
    ['result'] - the result of Domain.evaluate(). Typically includes:
        ['status'] - one of the STATUS_STRINGS
        ['loss'] - real-valued scalar that hyperopt is trying to minimize
    ['idxs'] - compressed representation of spec
    ['vals'] - compressed representation of spec
    ['tid'] - trial id (unique in Trials list)
```

Comment (s1dlx) [reply sd_webui_bayesian_merger/tpe_optimiser.py:58]: no, it's actually `best['result']['params']['base_alpha']`

File: sd_webui_bayesian_merger/tpe_optimiser.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            best['params']['weights'],
```
Comment (s1dlx) [inline sd_webui_bayesian_merger/tpe_optimiser.py:59]: `weights` doesn't exist, as each weight is an entry in the `best['result']['params']` dictionary.

A fix could be to parse the weights as

`best_weights = [best['result']['params'][f'block_{i}'] for i in range(NUM_TOTAL_BLOCKS)]`

and then use them in the call to `draw_unet`

**Review by s1dlx — CHANGES_REQUESTED** [review]
happy to merge after fixing the postprocessing function

**Review by s1dlx — APPROVED** [review]

**Pull request merged**